libvast/src/arrow_table_slice.cpp

```diff
@@ -1128,6 +1128,244 @@ arrow_table_slice<FlatBuffer>::record_batch() const noexcept {
+                    const auto [lhs_mismatch, rhs_mismatch]
+                      = std::mismatch(lhs.index.begin(), lhs.index.end(),
+                                      rhs.index.begin(), rhs.index.end());
+                    return lhs_mismatch == lhs.index.end();
+                  }),
+              "transformation indices must not be a subset of the following "
+              "transformation's index");
+  // The current unpacked layer of the transformation, i.e., the pieces required
+  // to re-assemble the current layer of both the record type and the record
+  // batch.
+  struct unpacked_layer {
+    std::vector<struct record_type::field> fields;
+    arrow::ArrayVector arrays;
+  };
+  const auto impl
+    = [](const auto& impl, unpacked_layer layer, offset index, auto& current,
+         const auto sentinel) noexcept -> unpacked_layer {
+    VAST_ASSERT(!index.empty());
+    auto result = unpacked_layer{};
+    // Iterate over the current layer. For every entry in the current layer, we
+    // need to do one of three things:
+    // 1. Apply the transformation if the index matches the transformation
+    //    index.
+    // 2. Recurse to the next layer if the index is a prefix of the
+    //    transformation index.
+    // 3. Leave the elements untouched.
+    for (; index.back() < layer.fields.size(); ++index.back()) {
+      const auto [is_prefix_match, is_exact_match]
+        = [&]() noexcept -> std::pair<bool, bool> {
+        if (current == sentinel)
+          return {false, false};
+        const auto [index_mismatch, current_index_mismatch]
+          = std::mismatch(index.begin(), index.end(), current->index.begin(),
+                          current->index.end());
+        const auto is_prefix_match = index_mismatch == index.end();
+        const auto is_exact_match
+          = is_prefix_match && current_index_mismatch == current->index.end();
+        return {is_prefix_match, is_exact_match};
+      }();
+      if (is_exact_match) {
+        VAST_ASSERT(current != sentinel);
+        for (auto&& [field, array] : std::invoke(
+               std::move(current->fun), std::move(layer.fields[index.back()]),
+               std::move(layer.arrays[index.back()]))) {
+          result.fields.push_back(std::move(field));
+          result.arrays.push_back(std::move(array));
+        }
+        ++current;
+      } else if (is_prefix_match) {
+        auto nested_layer = unpacked_layer{
+          .fields = {},
+          .arrays = caf::get<type_to_arrow_array_t<record_type>>(
+                      *layer.arrays[index.back()])
+                      .fields(),
+        };
+        nested_layer.fields.reserve(nested_layer.arrays.size());
+        for (auto&& [name, type] :
+             caf::get<record_type>(layer.fields[index.back()].type).fields())
+          nested_layer.fields.push_back({std::string{name}, type});
+        auto nested_index = index;
+        nested_index.push_back(0);
```

Comment:
Wdyt of renaming this `index` to `cursor`? That would be a little more precise imo.
`auto{index}.push_back(0)` would have been handy here as well.